lib/src/visitor/serialize.dart

```diff
@@ -1085,23 +1090,41 @@ class _SerializeVisitor
       return;
     }
 
-    _writeLineFeed();
+    bool indentNextChild;
+    if (_shouldWriteLineFeedBeforeChildren(children)) {
+      _writeLineFeed();
+      indentNextChild = true;
+    } else {
+      _writeOptionalSpace();
+      indentNextChild = false;
+    }
+
     CssNode previous;
-    _indent(() {
-      for (var i = 0; i < children.length; i++) {
-        var child = children[i];
-        if (_isInvisible(child)) continue;
+    for (var i = 0; i < children.length; i++) {
+      var child = children[i];
+      if (_isInvisible(child)) continue;
 
-        if (previous != null) {
-          if (_requiresSemicolon(previous)) _buffer.writeCharCode($semicolon);
+      if (previous != null) {
+        if (_requiresSemicolon(previous)) _buffer.writeCharCode($semicolon);
+        if (_shouldWriteLineFeedBeforeChild(child)) {
           _writeLineFeed();
-          if (previous.isGroupEnd) _writeLineFeed();
+          indentNextChild = true;
+        } else {
+          _writeOptionalSpace();
+          indentNextChild = false;
         }
-        previous = child;
+        if (previous.isGroupEnd) _writeLineFeed();
+      }
+      previous = child;
 
-        child.accept(this);
+      if (indentNextChild) {
+        _indentation++;
```

Comment:
I'm not sure I understand this indentation logic. It looks like if `indentNextChild` is false, it will print the child at one level of indentation lower than it would otherwise, but don't we want *zero* indentation for trailing comments?

I wonder if it makes more sense to completely special-case these comments, and write something more like:

```dart
if (_isTrailingComment(child)) {
  _writeOptionalSpace();
  _withoutIndentation(() => child.accept(this));
  continue;
}
```

For the post-bracket comments you'd want to do instead something like:

```dart
if (children.isNotEmpty && _isTrailingComment(children.first)) {
  _writeOptionalSpace();
  _withoutIndentation(() => children.first.accept(this));
  children = children.skip(1);
}
```

Reply:
I tried to incorporate your feedback here. PTAL.